Client setup

manoranjith opened the pull request:
<!-- Provide a general summary of your changes in the title above

Please read our contribution guidelines and sign the Contributor License
Agreement (CLA) before submitting the pull request. Also, check if there are no
other open pull requests targeting the same issue. -->

#### Description
<!-- Describe your changes in detail. -->
Added client initialization. The code has been organized into three packages: client, identity and ethereum.
1. **identity** package implements functionalities related to offchain identity and wallet management.
2. **client** package implements a client that combines the functionalities of channel controller client (from go-perun) and offchain network management (from identity package).
3. **ethereum** package isolates all imports that have direct or transient dependency to code from go-ethereum project, that is licensed under LGPL v3. All 
This package is structured such that, it can be moved out of this repository and can be converted from static linking to dynamic linking using the "[plugin package](https://golang.org/pkg/plugin/)". 
Additionally, a test package is added within ethereum, that provides pre-configured setups with a simulated ethereum backend for writing tests in other modules.

##### Category
<!-- Tell us what type of issue does your pull request target.
You can uncomment one of the following options: -->

<!-- Bug Fix -->
<!-- Improvement -->
Implementation Task

##### Relevant issue
<!-- Provide a link to the related issue. You can use the following keywords
and the issue number: "fixes", "resolves", "relates to". E.g.: closes #21

We accept only pull requests related to open issues. If you're suggesting a new
feature, improvement or fixing a bug that is not yet reported, please discuss it in
an issue before submitting a pull request. -->
Resolves #38 

#### Testing
<!-- Tell us how you have tested the changes. -->
Added unit tests.

##### Steps to run the tests
<!-- Describe a set of steps to run the tests relevant to this change. -->
1. Run the go test command from the root directory.

#### Checklist 
<!-- Please check if the pull request fulfils these requirements: -->

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

ggwpez commented on client/config.go:
```diff
@@ -0,0 +1,43 @@
+	ClientID string
+	User     identity.UserConfig
+
+	BC         BlockchainConfig
```
Maybe just `Blockchain` since you also called the other ones `User` and `Timeout`?

manoranjith replied:
Updated.

ggwpez commented on client/client_test.go:
```diff
@@ -0,0 +1,45 @@
+	})
+	t.Run("invalid-contracts", func(t *testing.T) {
+		rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+		s, cleanup := ethereumtest.NewContractBackendSetup(t, rng, "../testdata")
```
You can use the same setup in all three sub-tests since you only change it in the last one.

manoranjith replied:
Updated.

ggwpez commented on ethereum/contractBackend.go:
```diff
@@ -0,0 +1,97 @@
+// ContractBackend provides ethereum specific contract backend functionality.
+type ContractBackend struct {
+	// cb is the instance of contract backend that will be used for all onchain communications.
+	cb *ethchannel.ContractBackend
```
Can you embed or typedef this?

manoranjith replied:
I use `ethchannel.contractBackend` as a private member instead of embedding because the `ContractBackend` struct in this package implements the `ContractBackend` interface defined in `client/client.go`. And, I do not want this struct to expose the methods of the `ethchannel.ContractBackend`. The purpose of this interface is to make all the ethereum specific functions accessible using the abstract types defined in the core go-perun sdk.

Packages in `go-ethereum`, that are required in this project are licensed under LGPL v3 and it is one of the objectives to have a dst-go binary that does not contain any code licensed under LGPL v3.

By this arrangement all imports from `go-perun/backend/ethereum` and `go-ethereum` are restricted only to ethereum package. While the dst-go binary may still contain LGPL code as the `dst-go/ethereum` is statically linked with other packages, it is always possible (since we have proper abstractions defined) to covert the static linking to dynamic linking by using the `plugin` package in go, whenever we require a `dst-go` binary that is not containing any LGPL code.

ggwpez replied:
Im not familiar with the licence stuff, @sebastianst can have a look when he has time.

ggwpez commented on ethereum/contractBackend.go:
```diff
@@ -0,0 +1,97 @@
+
+// InitCustomContractBackend initializes a contract backend with the pre configured ethereum connection and wallet.
+// Use this function ONLY IN TEST PACKAGES to set a simulated backend and test wallets with weak encryption params.
+func InitCustomContractBackend(cb *ethchannel.ContractBackend) *ContractBackend {
```
This would be much better placed in `ethereum/test` or some other test package, otherwise it will be compiled into a release binary.

manoranjith replied:
Yeah, this is one thing that i wanted to change, but currently I am not finding a clean way to do this under these constraints:

1. Keep the properties of ContractBackend private, as it is an internal detail.
2. Setup a custom backend for testing, when tests are run in a separate package "ethereum_test" and not in the same package "ethereum". 

One way, i was thinking was to use build tags and move this function to a separate file, that is built  only during tests. But that implies "test" flag should be passed in the go test command everytime. 

Do you find any other better ways ?

ggwpez replied:
Looking at it again, this functions could just be called `NewContractBackend` and can then be used to create the `ContractBackend` that is passed to the client.  
The other `InitContractBackend` function can be privately put in the `[ethereum/test]` package. Just need to somehow expose the Scrypt-Params. This resolves the issue afaik.

manoranjith replied:
Yes, I have updated the name to `NewXYZ...` for both the backends. No, the function cannot return `*ethchannel.ContractBackend` because if the package is compiled as plugin and the symbol should be loaded, then the code that uses it will have to import `go-perun/ethereum/channel`. This in turn will cause the calling code to depend on `go-ethereum` library (transiently).
I have explained this in detail in the doc comments of the function as well as package docs.

manoranjith replied:
Regarding the other point, I figured out a way to do this using the `internal` package feature.

I moved all the implementation code to a separate package within the `internal` directory.  Code from this directory can imported only by packages in the internal's parent directory and its children. So even though the package has exported variables, it is not exposed outside of the ethereum package. However, with this approach I ended up with three packages, on each for real backend, test backend and implementation.

The only other way i saw to just use two packages (real backend and test backend), but that would cause a lot of code duplication. Hence i preferred the previous method.

ggwpez replied:
Ok great.

ggwpez commented on ethereum/contractBackend.go:
```diff
@@ -0,0 +1,97 @@
+
+// InitContractBackend initializes a connection to blockchain node and
+// sets up a wallet for funding onchain transactions.
+func InitContractBackend(url, userKeystorePath string, userAddr wallet.Address) (_ *ContractBackend, err error) {
```
Remove named returns.

manoranjith replied:
Updated.

ggwpez commented on ethereum/contractBackend.go:
```diff
@@ -0,0 +1,97 @@
+
+// InitContractBackend initializes a connection to blockchain node and
+// sets up a wallet for funding onchain transactions.
+func InitContractBackend(url, userKeystorePath string, userAddr wallet.Address) (_ *ContractBackend, err error) {
```
This function is also testing only?

manoranjith replied:
yes, see this [comment](https://github.com/direct-state-transfer/dst-go/pull/50#discussion_r444897627) above.

ggwpez commented on ethereum/contractBackend.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+// NewFunder initializes and returns an instance of ethereum funder.
+func (cb *ContractBackend) NewFunder(assetAddr wallet.Address) channel.Funder {
```
Why do you have these wrapper functions?

manoranjith replied:
See this [comment](https://github.com/direct-state-transfer/dst-go/pull/50#discussion_r443999203).

ggwpez commented on ethereum/contractBackend.go:
```diff
@@ -0,0 +1,97 @@
+	// Integrity of Adjudicator is implicitly done during validation of asset holder contract.
+	err := ethchannel.ValidateAssetHolderETH(ctx, cb, ethwallet.AsEthAddr(assetAddr), ethwallet.AsEthAddr(adjAddr))
+	if ethchannel.IsContractBytecodeError(err) {
+		return errors.New("invalid contracts at given address")
```
I think somewhere above you used `errors.Wrap`, so either use `Wrap` everywhere for errors where you dont know if they are already wrapped or not, or always use `WithMessage`.
In this line either `Wrap` or `WithMessage` since the original cause is lost otherwise.  
EDIT: Since you dont handle `ContractBytecodeError` errors differently, you dont need the `if`.

manoranjith replied:
Replaces `WithMessage` with `Wrap` when the stack trace is not yet captured.
I use a different error message for `ContractByteCodeError`, to differentiate between invalid byte code and any error in the process of validating the byte code.

ggwpez commented on ethereum/contractBackend_test.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+func Test_ContractBackend_Interface(t *testing.T) {
+	require.Implements(t, (*client.ContractBackend)(nil), new(ethereum.ContractBackend))
```
:laughing: i never saw this yet. Can you not just do it compile-time with  
`var _ client.ContractBackend = &ethereum.ContractBackend{}`?

manoranjith replied:
Yes, it also possible that way and is one the idiomatic ways of doing it.

Yet, i find this is a better approach, because the purpose is to test interface satisfaction and hence it fits better into  `_test.go`. Moreover, by moving it to test, the actual package need not import the `dst-go/client`. (Which it will do otherwise just for checking interface satisfaction).

ggwpez commented on ethereum/test/contractBackend.go:
```diff
@@ -0,0 +1,63 @@
+// NewContractBackendSetup initializes and returns a test ContractBackend.
+// It returns an initialized user with test accounts and a simulated contract backend,
+// with adjudicator and asset holder contracts deployed on it.
+func NewContractBackendSetup(t *testing.T, rng *rand.Rand, tempDir string) (ContractBackendSetup, func()) {
```
Is it important that the `tempDir` can be set or what is wrong with having it random?

manoranjith replied:
`tempDir` is the temporary directory in which temp data for testing will be created. In this case, it will be the `testdata` directory in the project root. 
I have updated the tests to use system's temp directory and hence this argument has been removed.

ggwpez commented on ethereum/test/contractBackend.go:
```diff
@@ -0,0 +1,63 @@
+// NewContractBackendSetup initializes and returns a test ContractBackend.
+// It returns an initialized user with test accounts and a simulated contract backend,
+// with adjudicator and asset holder contracts deployed on it.
+func NewContractBackendSetup(t *testing.T, rng *rand.Rand, tempDir string) (ContractBackendSetup, func()) {
```
Is this case, a naming for the `func()` return like `cleanup` would be nice.

manoranjith replied:
Updated.

ggwpez commented on ethereum/test/walletBackend.go:
```diff
@@ -0,0 +1,95 @@
+	User          *identity.User
+}
+
+func NewUserSetup(t *testing.T, rng *rand.Rand, tempDir string) (WalletBackendSetup, func()) {
```
Name the `cleanup` return.

manoranjith replied:
Updated.

ggwpez commented on ethereum/test/walletBackend.go:
```diff
@@ -0,0 +1,95 @@
+	Accounts      []wallet.Account
+}
+
+// NewWallet initializes a wallet with n accounts. Empty password string and weak encrytion parameters are used.
```
```suggestion
// NewWalletSetup initializes a wallet with n accounts. Empty password string and weak encrytion parameters are used.
```

manoranjith replied:
Updated.

ggwpez replied:
I think you `git` ate that line. It is still "`NewWallet`".

manoranjith replied:
yeah.. I it is updated now.

ggwpez commented on ethereum/test/walletBackend.go:
```diff
@@ -0,0 +1,95 @@
+		Keystore:      ws.Keystore,
+		KeystorePath:  ws.KeystorePath,
+		User:          user,
+	}, cleanup
```
Can you just return `walletCleanup` here?

manoranjith replied:
Updated.

ggwpez commented on ethereum/testdata/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
```
You just created the `testdata` directory for later, or why?

manoranjith replied:
Earlier I was using some fixed data for tests which were put in `testdata`. Then i rewrote the tests to not have any fixed data for tests. Hence `testdata` is not  required. I have removed it.

ggwpez commented on ethereum/walletBackend.go:
```diff
@@ -0,0 +1,60 @@
+
+// WalletBackend provides ethereum specific wallet backend functionality.
+type WalletBackend struct {
+	encParams scryptParams
```
Embed/typedef instead?

manoranjith replied:
In the same context as described in this [comment](https://github.com/direct-state-transfer/dst-go/pull/50#discussion_r443999203), `WalletBackend` type implements `WalletBackend` interface defined in identity package.

ggwpez commented on ethereum/walletBackend.go:
```diff
@@ -0,0 +1,60 @@
+// InitTestWalletBackend initializes an ethereum specific wallet backend with weak encryption parameters.
+// These wallets can be unlocked faster and intended for use in tests only.
+// For all other purposes, use the InitWalletBackend function instead.
+func InitTestWalletBackend() *WalletBackend {
```
Please make all testing functions either private or put them in `test` packages.

manoranjith replied:
yes, see this [comment](https://github.com/direct-state-transfer/dst-go/pull/50#discussion_r444897627) above.

manoranjith replied:
Updated. I have reorganized the packages as stated in [previous comment](https://github.com/direct-state-transfer/dst-go/pull/50#discussion_r451661152). So it is moved to `etherem/test` package.

ggwpez commented on ethereum/walletBackend.go:
```diff
@@ -0,0 +1,60 @@
+	return ethwallet.NewWallet(ks, password)
+}
+
+func (wb *WalletBackend) NewAccount(w wallet.Wallet, addr wallet.Address) (wallet.Account, error) {
```
Comment all public functions.

manoranjith replied:
Updated.

ggwpez commented on identity/config.go:
```diff
@@ -0,0 +1,20 @@
+type WalletConfig struct {
+	Addr         wallet.Address
+	KeystorePath string
+	Pwd          string //The passwords are received by the cli app as command line arguments are passed to the config.
```
Just write what it is, not how it is getting there. Also; `Pwd` on Unix normally means working directory.  
In the future this should probably not be a string, but the last time i searched for it, there was no good library for secure-strings in golang…

manoranjith replied:
removed the comment and updated the variable name to `Password`. 
Regarding secure way for handling passwords, can it be added as one the requirements in the [Identity Management task](#43) ?

ggwpez replied:
When it is a requirement from your side too, for sure. But i think it should go into #42 which is probably a sub-task of the identity managment.

ggwpez commented on identity/config.go:
```diff
@@ -0,0 +1,20 @@
+
+// WalletConfig defines wallet specific configuration parameters.
+type WalletConfig struct {
+	Addr         wallet.Address
```
What address is this?

manoranjith replied:
`WalletConfig` is used to specify the config for accessing any account. It can onchain account or perun id account. This address corresponds to whichever address is being used. See `UserConfig` to see how it is used.

ggwpez replied:
I just wondered because there is normally more than one `Account`/`Address` per Wallet. Looking at `UserConfig` there seem to be at most two Accounts and they are assumed to be in different wallets. 
In the future it is planned to have one off-chain-`Account` per `Channel`, so this abstraction would not work.  
The `demo-node` only does it as simplification.  Im not sure though if we should rework it now or leave it as is.

manoranjith replied:
Thanks for pointing out. My idea was use one wallet for on-chain account and another one for off-chain IDs and participant IDs (As mentioned in the comment).
But I missed to add a variable to keep track of participant IDs. I have add another variable `PartAdds` to the user that keeps track of all participant IDs for all the channels the user is participating in.

ggwpez commented on identity/config.go:
```diff
@@ -0,0 +1,20 @@
+type UserConfig struct {
+	Alias string
+
+	ListenerURL URL //Client will lister for incoming network connections at this URL
```
This is the endpoint for the perun-node, not the API, right? Otherwise one could think, it is the same endpoint for both.

manoranjith replied:
Yes, it is the endpoint for perun-node. I have renamed the variable to `NetworkAddress` to make it more clear.

ggwpez commented on identity/peer.go:
```diff
@@ -0,0 +1,30 @@
+type URL struct {
+	IPAddr   string
+	Port     uint16
+	Endpoint string
```
This is not used right now?

manoranjith replied:
It is used, but after thinking through i feel this struct may not handle all types of network address that we want to use. Since the scope of this issue is only to setup the client, i have updated the implementation to use string data type for URL. 

Later during the implementation of Identity Management module, we can define the requirements and implement the data type to handle URLs.

ggwpez commented on identity/peer.go:
```diff
@@ -0,0 +1,30 @@
+
+type Peer struct {
+	Alias          string
+	PerunID        wallet.Account
```
The `PerunID` is defined as `peer.Address`.

manoranjith replied:
Updated.

ggwpez commented on identity/user.go:
```diff
@@ -0,0 +1,56 @@
+	NewWallet(string, string) (wallet.Wallet, error)
+	NewAccount(wallet.Wallet, wallet.Address) (wallet.Account, error)
+}
+type User struct {
```
Newline

manoranjith replied:
Updated.

ggwpez commented on identity/user.go:
```diff
@@ -0,0 +1,56 @@
+
+	// Account for funding the channel and the onchain transactions.
+	OnChainAcc wallet.Account
+	// OnChainKeystorePath is the wallet that stores the keys for onchain account.
```
Wrong comment.

manoranjith replied:
Updated.

ggwpez commented on identity/user.go:
```diff
@@ -0,0 +1,56 @@
+	OnChainWallet wallet.Wallet
+
+	// Account corresponding to perunID.
+	// Used for authenticating user identity in offchain network.
```
Im not sure if this comment makes sense.

manoranjith replied:
Updated.

ggwpez commented on identity/user.go:
```diff
@@ -0,0 +1,56 @@
+// unlocks both onchain, perun account with indefinite timeout.
+func NewUnlockedUser(wb WalletBackend, cfg UserConfig) (User, error) {
+	u := User{}
+	var err error
```
Define both as named return values.

manoranjith replied:
Updated.

ggwpez replied:
Now there is a newline after the function starts.

ggwpez commented on identity/user.go:
```diff
@@ -0,0 +1,56 @@
+	NewWallet(string, string) (wallet.Wallet, error)
+	NewAccount(wallet.Wallet, wallet.Address) (wallet.Account, error)
+}
+type User struct {
```
Comment `User`.

manoranjith replied:
Updated.

ggwpez commented on identity/user_test.go:
```diff
@@ -0,0 +1,98 @@
+						KeystorePath: setup.KeystorePath,
+						Pwd:          ""},
+					OffChainWallet: identity.WalletConfig{
+						Addr:         setup.User.OnChainAcc.Address(),
```
```suggestion
						Addr:         setup.User.OffChainAcc.Address(),
```  
same below.

manoranjith replied:
Updated.

ggwpez commented on identity/user_test.go:
```diff
@@ -0,0 +1,98 @@
+					OnChainWallet: identity.WalletConfig{
+						Addr:         setup.User.OnChainAcc.Address(),
+						KeystorePath: setup.KeystorePath,
+						Pwd:          "invalid-pwd"}}},
```
This does not fail when i change the password to `""`.

manoranjith replied:
Thanks for testing it out. It fails because `OffchainWallet` is not defined. Have fixed the test, not it should fail with correct password.

ggwpez commented on identity/user_test.go:
```diff
@@ -0,0 +1,98 @@
+					Alias: setup.User.Alias,
+					OnChainWallet: identity.WalletConfig{
+						Addr:         setup.User.OnChainAcc.Address(),
+						KeystorePath: "invalid-keystore-path",
```
Why does this work when i replace it with the correct `setup.KeystorePath`?

manoranjith replied:
Thanks for testing it out. It fails because `OffchainWallet` is not defined. Have fixed the test, not it should fail with correct keystore.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,109 @@
+// for transacting on these channels.
+func InitClient(blockchain ContractBackend, user identity.User, adj, assetHolder wallet.Address) (*Client, error) {
+	perunLogrus.Set(logrus.DebugLevel, logrus.StandardLogger().Formatter)
+	dialTimeout := time.Second * nodeConfig.Timeout.Response
```
timeout.Response is a variable of type `time.Duration`, so this is a mistake. Fixed.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,109 @@
+// This method is called on every incoming channel proposal.
+// TODO: Replace the AcceptAll Handler with a proper implementation once RPC module is implemented.
+func (c *Client) HandleProposal(_ *client.ChannelProposal, responder *client.ProposalResponder) {
+	ctx, cancel := context.WithTimeout(context.Background(), nodeConfig.Timeout.Response)
```
Yes, it has been fixed in the review commit.

manoranjith commented on client/client_test.go:
```diff
@@ -0,0 +1,45 @@
+	"github.com/stretchr/testify/require"
+	_ "perun.network/go-perun/backend/ethereum/wallet/test"
+
+	"github.com/direct-state-transfer/dst-go/client"
```
Updated.

manoranjith commented on client/client_test.go:
```diff
@@ -0,0 +1,45 @@
+	ethereumtest "github.com/direct-state-transfer/dst-go/ethereum/test"
+
+	"github.com/stretchr/testify/require"
+	_ "perun.network/go-perun/backend/ethereum/wallet/test"
```
No it is not needed now. I have removed it.

manoranjith commented on client/config.go:
```diff
@@ -0,0 +1,43 @@
+
+// TimeoutConfig defines the different timeout configurations.
+type TimeoutConfig struct {
+	//Maximum duration the node will wait until the channel is funded by all participants.
```
Have updated all comments to have a space after `//`.

manoranjith commented on client/config.go:
```diff
@@ -0,0 +1,43 @@
+// TimeoutConfig defines the different timeout configurations.
+type TimeoutConfig struct {
+	//Maximum duration the node will wait until the channel is funded by all participants.
+	//If it expires, channel will be closed.
```
Updated the comment.

manoranjith commented on client/config.go:
```diff
@@ -0,0 +1,43 @@
+	Funding time.Duration
+	//Maximum duration a peer or user can take for responding to requests.
+	//If it expires, the peer or user will be considered offline.
+	Response time.Duration
```
I have updated the timeout used in `HandleProposal` to meet the requirements of the `Accept` function, pointed out in the link.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,109 @@
+// registering new peers on the network, proposing channels with the peers, and
+// for transacting on these channels.
+func InitClient(blockchain ContractBackend, user identity.User, adj, assetHolder wallet.Address) (*Client, error) {
+	perunLogrus.Set(logrus.DebugLevel, logrus.StandardLogger().Formatter)
```
Updated.

ggwpez commented:
I gave a :+1: to the resolved discussions and left the ones where i was unsure or waiting for Sebastian open.

manoranjith commented:
Thanks for the review comments (@ggwpez). I think I have most of them, except a few open points.

- Now I have restructured the project to provide a proper package organization (much of the code still remains the same) and added detailed doc comments.
- This current package organization aims to make the project more readable, have loosely coupled components and more testable. I will add a proposal referencing to this issue, in which I can describe the structure in detail and how it works and why it is better.
- Hence, I have removed the `WIP` tag and opened the PR for complete review.

manoranjith commented:
I have added proposal [here](https://github.com/direct-state-transfer/dst-proposal/pull/2).

anagha-ks commented on client/client.go:
```diff
@@ -0,0 +1,141 @@
+// NewEthereumPaymentClient initializes a two party, ethereum payment channel client for the given user.
+// It establishes a connection to the blockchain and verifies the integrity of contracts at the given address.
+// It uses the comm backend to initialize adapters for off-chain communication network.
+func NewEthereumPaymentClient(cfg Config, user dst.User, comm dst.CommBackend) (*Client, error) {
```
Can we make the function NewEthereumPaymentClient function into seperate purpose specific small functions ?
one each for validating the integrity of contracts, establishing connection to blockchain and initializing adapters for off-chain communication network?

manoranjith replied:
Updated.

anagha-ks commented on client/config.go:
```diff
@@ -0,0 +1,21 @@
+	ChainConnTimeout time.Duration
+
+	// Path to directory containing persistence database
+	PersistenceDir string
```
Is it better to use DatabaseDir or something related to DB and Dir, like PersistenceDbDir  instead of PersistenceDir, so that from the name itself we can catch it is a database directory path.?

manoranjith replied:
I have updated the variable name to `DatabaseDir`.

ggwpez commented on blockchain/ethereum/internal/wallet.go:
```diff
@@ -0,0 +1,78 @@
+	return ethwallet.NewWallet(ks, password)
+}
+
+// NewAccount retreives the account correspoding to the given address, unlocks and returns it.
```
```suggestion
// NewAccount retrieves the account corresponding to the given address, unlocks and returns it.
```

manoranjith replied:
Updated.

ggwpez commented on blockchain/ethereum/internal/wallet.go:
```diff
@@ -0,0 +1,78 @@
+	addr := ethwallet.AsWalletAddr(common.HexToAddress(str))
+
+	zeroValue := ethwallet.Address{}
+	if addr.Equals(&zeroValue) && str != zeroValue.String() {
```
This ready a bit weird. You can also return `ethwallet.Address{}` in case that `str` is empty and error otherwise when it encounters the `zeroValue`.

manoranjith replied:
Yeah, it was a bit tricky because `common.HexToAddress` parses even invalid strings to zero value of the address type.
So, this check makes the function to return an error when addr has zero value and the input string is not a valid zero value representation of the address. 

As an illustration, without this check "invalid-string" would be parsed to zero value of address type, while it is an invalid address.

As you can find in the corresponding test, this can now
1.  Parse a valid non-zero address string without error.
2.  Parse a valid zero address string without error.
3. Error on invalid inputs.

I will also add a brief explanation as inline comments in code as well.

manoranjith replied:
Added a comment in the code.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ethereumtest
```
It is named `ethereumtest` because there can be multiple test helper packages in the project and it is good to have the package name unique within the project.

Also,
- When a test uses helpers from two packages, say `ethereum/ethereumtest` and `node/nodetest`, no renaming of imported packages is required.
- `ethereumtest.NewChainBackendSetup` is more explicit in its meaning compared to `test.NewChainBackendSetup`. Because there will be more than one test packages, if it is named the other way.

manoranjith replied:
Also from the [tests in go-ethereum](https://github.com/ethereum/go-ethereum/blob/master/common/types_test.go#L104), `0x0000000000000000000000000000000000000000` is a valid address representation.

manoranjith commented on blockchain/ethereum/ethereumtest/wallet.go:
```diff
@@ -0,0 +1,93 @@
+// NewRandomAddress generates a random wallet address. It generates the address only as a byte array.
+// Hence it does not generate any public or private keys corresponding to the address.
+// If you need an address with keys, use Wallet.NewAccount method.
+func NewRandomAddress(rnd *rand.Rand) wallet.Address {
```
Yes, I find it the same. But i was not importing the `go-perun/backend/ethereum/wallet/test` for any other functionality. So i felt it was simpler to define it in 3 lines of code it in this package instead of importing another package.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+	"testing"
+	"time"
+
+	"github.com/direct-state-transfer/dst-go/blockchain/ethereum/internal"
```
Updated.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+	AdjAddr, AssetAddr wallet.Address
+}
+
+// NewChainBackendSetup returns a simulated contract backend  with assetHolder, adjudicator contracts deployed.
```
Updated.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// NewChainBackendSetup returns a simulated contract backend  with assetHolder, adjudicator contracts deployed.
+// It also generate the given number of accounts and funds them each with 10 ether.
```
Updated.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+// It also generate the given number of accounts and funds them each with 10 ether.
+// and returns a test ChainBackend using the given randomness.
+func NewChainBackendSetup(t *testing.T, rng *rand.Rand, cntFundedAccs uint) (_ *ChainBackendSetup) {
+	// userSetup := NewUserSetup(t, rng)
```
Yeah, i have removed the commented out line of code.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+// NewChainBackendSetup returns a simulated contract backend  with assetHolder, adjudicator contracts deployed.
+// It also generate the given number of accounts and funds them each with 10 ether.
+// and returns a test ChainBackend using the given randomness.
+func NewChainBackendSetup(t *testing.T, rng *rand.Rand, cntFundedAccs uint) (_ *ChainBackendSetup) {
```
Updated.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultTxTimeout)
+	defer cancel()
+	for i := range accs {
+		simBackend.FundAddress(ctx, ethwallet.AsEthAddr(accs[i].Address()))
```
Updated.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -0,0 +1,88 @@
+
+	onChainAcc := &accs[0].(*ethwallet.Account).Account
+	contractBackend := ethchannel.NewContractBackend(simBackend, ks, onChainAcc)
+	return contractBackend
```
Updated.

manoranjith commented on blockchain/ethereum/ethereumtest/wallet.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// WalletSetup can generate any number of keys for testing. To enable faster unlocking of keys, it uses
+// weak encryption parameters for storage encryption of keys .
```
Updated, used "the keys" instead of "the the keys". ;)

manoranjith commented on blockchain/ethereum/ethereumtest/wallet.go:
```diff
@@ -0,0 +1,93 @@
+
+	t.Cleanup(func() {
+		err := os.RemoveAll(ksPath)
+		if err != nil {
```
Updated.

manoranjith commented on blockchain/ethereum/internal/chain_test.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func Test_ChainBackend_Deploy(t *testing.T) {
+	rng := rand.New(rand.NewSource(1729))
```
Yeah, it looks interesting. But I would create another issue to add this functionality in dst-go.

manoranjith commented on blockchain/ethereum/internal/chain.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+func validateContracts(cb ethchannel.ContractBackend, adjAddr, assetAddr wallet.Address) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
Yes, thanks for spotting the error. I have updated the implementation to use a package level constant for on-chain transaction timeout.

ggwpez commented on blockchain/ethereum/internal/wallet_test.go:
```diff
@@ -0,0 +1,100 @@
+	rng := rand.New(rand.NewSource(1729))
+	wb := ethereumtest.NewTestWalletBackend()
+
+	t.Run("happy_non_zero_value", func(t *testing.T) {
```
Please also add some hardcoded test values like `123`, `0xbd5465321`, `0x931d387731bbbc988b312206c74f77d004d6b84b`, `931d387731bbbc988b312206c74f77d004d6b84b` or `c56432f` to test the format and length requirements.

manoranjith replied:
Thanks for the suggestion, I found out that zero value string can be of different length `""` or `"0x"` or `"0x000"` or the canonical form "0x" followed by 40 zeros.

I have added tests for such data and also updated the implementation to handle this.

ggwpez commented on blockchain/ethereum/internal/chain.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+// NewFunder initializes and returns an instance of ethereum funder.
+func (cb *ChainBackend) NewFunder(assetAddr wallet.Address) channel.Funder {
```
`NewFunder` and `NewAdjudicator` are untested right now.

manoranjith replied:
Yes, I am not sure what to test. Because, the only expected behavior of these functions is that they return valid interfaces. 

Due to special structuring (as described in the package docs) of this package, all its method signature should not contain types defined in `go-ethereum` or `go-perun/backend/ethereum`, the return type is already an interface.

If it was otherwise returning a struct type, i would have added test to check if the structs implement corresponding interfaces.

Do you have suggestion for testing these two APIs ?

ggwpez commented on client/client.go:
```diff
@@ -0,0 +1,159 @@
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/peer"
+
+	"github.com/pkg/errors"
```
Import grouping.

manoranjith replied:
Updated.

manoranjith commented on blockchain/ethereum/internal/wallet.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// ParseAddr parses the ethereum address from the given string.
+func (wb *WalletBackend) ParseAddr(str string) (wallet.Address, error) {
```
Adding a `0x` prefix is optional. Have updated the doc comments.

manoranjith commented on blockchain/ethereum/internal/wallet_test.go:
```diff
@@ -0,0 +1,100 @@
+		gotAddr, err := wb.ParseAddr(validAddr.String())
+		assert.NoError(t, err)
+		require.NotNil(t, gotAddr)
+		assert.Equal(t, validAddr.Bytes(), gotAddr.Bytes())
```
Updated.

manoranjith commented on blockchain/ethereum/internal/wallet_test.go:
```diff
@@ -0,0 +1,100 @@
+	setup := ethereumtest.NewWalletSetup(t, rng, 1)
+
+	t.Run("happy", func(t *testing.T) {
+		w, err := wb.NewAccount(setup.Wallet, setup.Accs[0].Address())
```
Yes, it can be called multiple times. Add a test to cover this.

manoranjith commented on blockchain/ethereum/internal/wallet.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// NewAccount retrieves the account corresponding to the given address, unlocks and returns it.
+func (wb *WalletBackend) NewAccount(w wallet.Wallet, addr wallet.Address) (wallet.Account, error) {
```
Changed to 'UnlockAccount`.

ggwpez commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dst defines domain types and interfaces used in this project.
```
```suggestion
```
already in `doc.go`.

manoranjith replied:
Updated.

ggwpez commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+	NewDialer() peer.Dialer
+}
+
+// Credential represents the parameters required for a to create a signature for a given address.
```
```suggestion
// Credential represents the parameters required for a ??? to create a signature for a given address.
```

manoranjith replied:
Updated.

ggwpez commented on node/doc.go:
```diff
@@ -0,0 +1,19 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package node implements the an node that can run multiple instances of
```
```suggestion
// Package node implements a node that can run multiple instances of
```
i do not find the *node* that it speaks of btw.

manoranjith replied:
The node itself will be implemented in the later tasks. The scope of this pull request is initialize the client, identify the necessary components and provide an initial structure to the package. 

Currently, only user initialization functionality is implemented in this package.

ggwpez commented on node/user.go:
```diff
@@ -0,0 +1,83 @@
+// NewUnlockedUser initializes a user and unlocks all the accounts.
+// (those corresponding to on-chain address, off-chain address and all participant addresses.
+func NewUnlockedUser(wb dst.WalletBackend, cfg UserConfig) (dst.User, error) {
+	if wb == nil {
```
Also check `cfg` if you need input validation.

manoranjith replied:
Input validation is not needed for `cfg`, as it will not cause the program to panic. I have added a test to ensure this.

ggwpez commented on node/user.go:
```diff
@@ -0,0 +1,83 @@
+	if wb == nil {
+		return dst.User{}, errors.New("wallet backend should not be nil")
+	}
+	var err error
```
Use named returns.

manoranjith replied:
I find named returns useful only when it adds more meaning to the function signature. Do you feel, having named returns adds more meaning ?

I also find this is one of the common good practices from the [google code review comments](https://gist.github.com/adamveld12/c0d9f0d5f0e1fba1e551#named-result-parameters).

ggwpez replied:
I normally just do it if it makes the function shorter :laughing:

manoranjith replied:
Yeah, I was also doing the same. :smiley:

But I found it appealing that, doing it this way makes it look better in the docs, so adopted it. :wink:

ggwpez commented on node/user.go:
```diff
@@ -0,0 +1,83 @@
+
+// parseUnlock parses the given addresses string using the wallet backend and unlocks accounts
+// corresponding to each of the given addresses.
+func parseUnlock(wb dst.WalletBackend, w wallet.Wallet, addrs ...string) ([]wallet.Address, error) {
```
You are only using this function with multiple addresses once anyway, so i think it is superfluous. It has also more than one responsibility which is unclean AFAIK.

manoranjith replied:
Yes, I use this function twice: once with a single address and once with multiple addresses, so i feel it is makes the calling code more simpler, than when having a single address and calling this function in a for loop.

Also regarding the responsibility,  this function parses and unlocks the wallet. I don't think there are need to be defined as separate functions because they don't implement those functionalities but invoke the corresponding functions on the provided wallet backend.

ggwpez commented on node/user.go:
```diff
@@ -0,0 +1,83 @@
+	return u, nil
+}
+
+// newWalletAcc initializes the wallet using the wallet backend and unlocks accounts corresponding
```
```suggestion
// newWallet initializes the wallet using the wallet backend and unlocks accounts corresponding
```
>to each of the given addresses

There is only one given

manoranjith replied:
Updated.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func initComm(comm dst.CommBackend, listenAddr string) (peer.Dialer, peer.Listener, error) {
+	if comm == nil {
```
Only `comm` it checked because it it is nil, its usage in the following lines will cause the program to panic and shutdown abruptly. 

I do it here, instead of the `NewEthereumPaymentClient` because the usage of the interface is here and it is better to have the nil validation closer to the usage.

manoranjith replied:
I would also not use panic, because it will cause the node to stop abruptly, it is better to have a gracious shutdown in case of errors.

manoranjith replied:
When giving it another thought, I think, input validation (for nil value of interfaces) is completely not required in this package. Because values for these interfaces will be initialized and injected by the node package. 

if all the validation with respect to interfaces are implemented in the node package, then it would be safe to assume the interface arguments in the functions off all component packages will be supplied with non nil values. As a consequence, the node will be panic due to a nil interface in the component package.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+func (c *Client) Close() error {
+	err := c.Client.Close()
+	c.wg.Wait()
+	return errors.Wrap(err, "closing channel client")
```
I have added the stack trace here, because `client.Close` is function from an external library. It is safer to make no expectation from the external library regarding whether or not stack trace has been attached.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+	}
+	err = chain.ValidateContracts(adjudicatorAddr, assetAddr)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "validating contracts")
```
Nope, i add stack trace at the point when a new error is created or when a function from an external library is called. It is not needed in this case.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+	// will be unlocked by the client when required.
+	offChainAcc, err := user.OffChain.Wallet.Unlock(user.OffChain.Addr)
+	if err != nil {
+		return nil, errors.WithMessage(err, "off-chain address")
```
I have revisited all the error messages and structured it such that, additional message is added only when some meaningful information can be added.

In this particular case, the `Unlock` function will add a message  "unlocking" and at this point, only "off-chain address" will be added. So that, the complete error message would be  "off-chain address: unlocking".

manoranjith commented on node/user.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+// NewUnlockedUser initializes a user and unlocks all the accounts.
+// (those corresponding to on-chain address, off-chain address and all participant addresses.
```
Updated.

manoranjith commented on blockchain/ethereum/internal/wallet_test.go:
```diff
@@ -0,0 +1,135 @@
+		for _, test := range tests {
+			t.Run(test.name, func(t *testing.T) {
+				gotAddr, err := wb.ParseAddr(test.input)
+				_, _ = gotAddr, err
```
Updated.

ggwpez commented on client/client_test.go:
```diff
@@ -0,0 +1,185 @@
+	// --account="0x6aeeb7f09e757baa9d3935a042c3d0d46a2eda19e9b676283dce4eaf32e29dc9,100000000000000000000000"
+	//
+	// Deploy asset holder and adjudicator contracts.A quick hacky solution is to clone the project at
+	// github.com/perun-network/perun-eth-demo, build it and run the below cmd:
```
```suggestion
	// github.com/perun-network/perun-eth-demo, checkout out dccb8d7, build it and run the below cmd:
```  
we changed it to `Alice`, so better fix it to a specific version.

manoranjith replied:
In the comments, I have added the git revision of the code to be used.

manoranjith replied:
I have added setup functions to deploy the contracts. You can find it in [`integration_setup_test.go`](https://github.com/direct-state-transfer/dst-go/pull/50/commits/4babc40b93da3e6c2d770d3085e51154916a8f89#diff-9620ed9af0d0854b0730c37b41c8f6de). Hence I have also removed the reference to `perun-eth-demo`.

ggwpez replied:
Yes this is much better.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+	walletBackend := ethereum.NewWalletBackend()
+	assetAddr, err := walletBackend.ParseAddr(cfg.Asset)
+	if err != nil {
+		return nil, nil, errors.WithMessage(err, "asset address")
```
`ParseAddr` function will add the message "parsing error" and at this point, i only added what was being parsed.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+	}
+	adjudicatorAddr, err := walletBackend.ParseAddr(cfg.Adjudicator)
+	if err != nil {
+		return nil, nil, errors.WithMessage(err, "adjudicator address")
```
`ParseAddr` function will add the message "parsing error" and at this point, i only added what was being parsed.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+func (c *Client) runAsGoRoutine(f func()) {
+	c.wg.Add(1)
+	go func(wg *sync.WaitGroup) {
+		f()
```
Updated.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+		return nil, nil, err
+	}
+	err = chain.ValidateContracts(adjudicatorAddr, assetAddr)
+	if err != nil {
```
Yes, I think it is better to directly return whenever possible. Have updated.

manoranjith commented on client/config.go:
```diff
@@ -0,0 +1,42 @@
+type Config struct {
+	Chain ChainConfig
+
+	// Path to directory containing persistence database
```
Updated.

manoranjith commented on client/client_test.go:
```diff
@@ -0,0 +1,192 @@
+// +build integration
```
Updated.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+
+	// List of participant addresses for this user in each open channel.
+	// OffChain credential is used for managing all these accounts.
+	PartAddrs []wallet.Address
```
The first implementation would use same address for all channels, quickly this will be followed by the update to use ephemeral keys for each channels. This field such ephemeral addresses. The keys corresponding to those addresses can be accessed using the same credentials as the off-chain wallet.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+	NewAdjudicator(adjAddr, receiverAddr wallet.Address) channel.Adjudicator
+}
+
+// WalletBackend wraps the methods for instanting wallets and accounts that are specific to a blockchain platform.
```
Updated.

manoranjith commented on node/config.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// UserConfig defines the parameters required to configure a user.
+// Addresss corresponding the respective accounts are stored as strings
```
I have updated the comment as below:

// UserConfig defines the parameters required to configure a user.
// Address strings should be parsed using the wallet backend.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+	Log() perunLog.Logger
+}
+
+// ChainBackend wraps the methods required for instanting and using components for
```
Updated.

manoranjith commented on client/client.go:
```diff
@@ -0,0 +1,157 @@
+// This method is called on every incoming channel proposal.
+// TODO: Implement an accept all handler until user api components are implemented.
+// TODO: Replace with proper implementation after user api components are implemented.
+func (ph *proposalHandler) HandleProposal(_ *client.ChannelProposal, _ *client.ProposalResponder) {}
```
Updated.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+
+	OffChainAddr peer.Address // Permanent identity used for authenticating the peer in the off-chain network.
+
+	CommAddr string // Address for off-chain communication
```
Updated.

manoranjith commented on comm/tcp/tcp.go:
```diff
@@ -0,0 +1,57 @@
+
+// Backend is an off-chain communication backend that implements adapters for
+// for tcp protocol. It stores configuration required for initializing the adapters.
+type Backend struct {
```
The primary purpose of `Backend` is to implement the necessary functionalities for using a particular communication protocol in the off-chain network such as TCP, Websockets, MQTT etc.,

It additionally holds any properties that is required.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+	OffChainAddr peer.Address // Permanent identity used for authenticating the peer in the off-chain network.
+
+	CommAddr string // Address for off-chain communication
+	CommType string // Type of off-chain communication protocol
```
Currently this is just a string. The possible values and how to define them will be added in later tasks. The scope of this pull request is just to initialize the state channel client and provide an initial structure to the project by identifying the different components required for initialization.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+//go:generate mockery -name CommBackend -output ./internal/mocks
+
+// CommBackend defines the set of methods required for initializing components required for off-chain communication.
+type CommBackend interface {
```
The `CommBackend` provides an abstraction to use different implementations for off-chain communication protocol adapter. Different implementations may require different kind of parameters for initializing the dialer and listener. 

By having a single interface, it is possible to add these parameters when initializing the backend and just call `NewDialer` and `NewListener` with the given signatures.

manoranjith replied:
Also, I find these two functions cohesive enough that it is better to have one interface the clearly describes the communication backend than having two.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+// the same type and version of communication and state channel protocol. If a user desires to use multiple types or
+// versions of any protocol, it should request a seprate session for each combination of type and version of those.
+type Session struct {
+	ID   string // ID is the unique identifier each instance of session.
```
Updated.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// ChannelClient allows the user to establish off-chain channels and transact on these channels.
+// A channel is closed when a close is initiated either by the user or other channel participants in the channel.
```
Yes, I also agree. i have removed the sentence. I feel it is sufficient to have only a high level description.

manoranjith commented on dst.go:
```diff
@@ -0,0 +1,131 @@
+
+// CommBackend defines the set of methods required for initializing components required for off-chain communication.
+type CommBackend interface {
+	// Returns a listener that can listen incommig messages at the specified address using the communication protocol.
```
Updated. The interface itself allows implementations for different protocols to be used. I have added some examples (TCP, Websockets, MQTT) in the comments.

ggwpez commented on comm/tcp/tcp_test.go:
```diff
@@ -0,0 +1,57 @@
+	listener, err := backend.NewListener(listenerAddr)
+	defer func() {
+		err := listener.Close() //nolint:govet 	// does not shadow the error defined at function level.
+		if err != nil {
```
```suggestion
		if err := listener.Close(); err != nil {
```

manoranjith replied:
I have updated this and few other places in the project wherever it was possible.
